Open Judge System/Services/OJS.Services/RedisCacheService.cs

```diff
@@ -0,0 +1,248 @@
+                if (!this.MemoryCacheHasKey(exTypeAsString, ex.Message))
+                {
+                    await this.emailSenderService.SendEmailAsync(this.devEmail, exTypeAsString, ex.Message);
+                }
+            }
+        }
+
+        public void Remove(string cacheId)
+        {
+            try
+            {
+                this.redisCache.KeyDelete(cacheId);
+            }
+            catch (RedisConnectionException ex)
+            {
+                var exTypeAsString = ex.GetType().ToString();
+                if (!this.MemoryCacheHasKey(exTypeAsString, ex.Message))
+                {
+                    this.emailSenderService.SendEmail(this.devEmail, exTypeAsString, ex.Message);
+                }
+            }
+        }
+
+        public async Task RemoveAsync(string cacheId)
+        {
+            try
+            {
+                await this.redisCache.KeyDeleteAsync(cacheId);
+            }
+            catch (RedisConnectionException ex)
+            {
+                var exTypeAsString = ex.GetType().ToString();
+                if (!this.MemoryCacheHasKey(exTypeAsString, ex.Message))
+                {
+                    await this.emailSenderService.SendEmailAsync(this.devEmail, exTypeAsString, ex.Message);
+                }
+            }
+        }
+
+        private async Task VerifyValueInCacheAsync<T>(
+            string cacheId,
+            Func<Task<T>> getItemCallback,
+            TimeSpan? expiration)
+        {
+            if (!await this.redisCache.KeyExistsAsync(cacheId))
+            {
+                var result = await getItemCallback();
+
+                var parsedValue = JsonConvert.SerializeObject(result);
+                await this.redisCache.StringSetAsync(
+                    cacheId,
+                    parsedValue,
+                    expiration);
+            }
+        }
+
+        private void VerifyValueInCache<T>(
+            string cacheId,
+            Func<T> getItemCallback,
+            TimeSpan? expiration)
+        {
+            if (!this.redisCache.KeyExists(cacheId))
+            {
+                var result = getItemCallback();
+
+                var parsedValue = JsonConvert.SerializeObject(result);
+                this.redisCache.StringSet(
+                   cacheId,
+                   parsedValue,
+                   expiration);
+            }
+        }
+
+        private bool MemoryCacheHasKey(string key, string value)
+        {
+            if (HttpRuntime.Cache.Get(key) == null)
+            {
+                HttpContext.Current.Cache.Add(
+                   key,
+                   value,
+                   null,
+                   DateTime.Now.AddMinutes(this.memmoryCacheExpirationInMinutes),
+                   TimeSpan.Zero,
+                   CacheItemPriority.Default,
+                   null);
+                return false;
+            }
+            return true;
```

Comment:
Is misleading to return `false` when the key is already in the cache. Either rename the method or refactor the logic, so the Set/Add is separate from the check.

Reply:
It was renamed